libs/muddle/src/punishment_broadcast_channel.cpp

```diff
@@ -0,0 +1,252 @@
+{
+  FETCH_UNUSED(ordered_delivery);
+  this->Expose(PULL_INFO_FROM_PEER, this, &PunishmentBroadcastChannel::AllowPeerPull);
+
+  // TODO(HUT): rpc beacon rename.
+  // Attaching the protocol
+  rpc_server_ = std::make_shared<Server>(endpoint_, SERVICE_PBC, channel_);
+  rpc_server_->Add(RPC_BEACON, this);
+
+  // Connect states to the state machine
+  state_machine_->RegisterHandler(State::INIT, this, &PunishmentBroadcastChannel::OnInit);
+  state_machine_->RegisterHandler(State::RESOLVE_PROMISES, this,
+                                  &PunishmentBroadcastChannel::OnResolvePromises);
+}
+
+PunishmentBroadcastChannel::~PunishmentBroadcastChannel() = default;
+//{
+//  // TODO(HUT): reinstate this fix.
+//  /* rpc_server_->Remove(RPC_BEACON); */
+//}
+
+// Interface methods
+bool PunishmentBroadcastChannel::ResetCabinet(CabinetMembers const &cabinet)
+{
+  FETCH_LOCK(lock_);
+  current_cabinet_ = cabinet;
+
+  // Set threshold depending on size of committee
+  if (current_cabinet_.size() % 3 == 0)
+  {
+    threshold_ = static_cast<uint32_t>(current_cabinet_.size() / 3 - 1);
+  }
+  else
+  {
+    threshold_ = static_cast<uint32_t>(current_cabinet_.size() / 3);
+  }
+
+  assert(current_cabinet_.size() > 3 * threshold_);
+
+  return true;
+}
+
+/**
+ * Set/reset the question, saving the old question, so that peer who have not yet
+ * reset can still access this information
+ *
+ */
+void PunishmentBroadcastChannel::SetQuestion(ConstByteArray const &question,
+                                             ConstByteArray const &answer)
+{
+  FETCH_LOCK(lock_);
+  previous_question_ = std::move(question_);
+  question_          = QuestionStruct(question, answer, certificate_, current_cabinet_);
+}
+
+void PunishmentBroadcastChannel::SetQuestion(std::string const &question,
+                                             std::string const &answer) /* */
+{
+  SetQuestion(ConstByteArray(question), ConstByteArray(answer));
+}
+
+void PunishmentBroadcastChannel::Enable(bool enable)
+{
+  FETCH_LOCK(lock_);
+  enabled_ = enable;
+}
+
+/**
+ * The channel needs to be driven to resolve network events
+ */
+std::weak_ptr<fetch::core::Runnable> PunishmentBroadcastChannel::GetRunnable()
+{
+  return state_machine_;
+}
+
+/**
+ * Determine whether the current question has been answered by all peers.
+ * If not, continue to try to resolve answers
+ */
+bool PunishmentBroadcastChannel::AnsweredQuestion()
+{
+  uint32_t messages_confirmed = 0;
+
+  for (auto const &node_plus_answers : question_.table_)
+  {
+    if (QuestionStruct::GetSeen(node_plus_answers.second).size() >= threshold_ &&
+        node_plus_answers.first != question_.self_)
+    {
+      messages_confirmed++;
+    }
+  }
+
+  bool const answered_question = messages_confirmed == question_.cabinet_.size() - 1;
+
+  return answered_question;
+}
+
+PunishmentBroadcastChannel::State PunishmentBroadcastChannel::OnInit()
+{
+  // Determine whether to take action
+  {
+    FETCH_LOCK(lock_);
```

Comment:
This lock only protects this small block, right? (And below, in `OnResolvePromises`.)

Reply:
Also importantly ```AllowPeerPull```

Reply:
I mean, other parts of this function, don't they need a lock?

Reply:
Oh - no, I have written it in such a way (I believe) that it should be thread safe, only locking critical sections. I ran it with thread sanitizer on so I am reasonably confident of this, unless I have accidentally changed something in the meantime

Reply:
Ok.